provider/slack/index.ts

```diff
@@ -49,46 +54,236 @@ async function listAllChannels(client: WebClient): Promise<ChannelInfo[]> {
+                relatedInfo: attachmentInfo.title || '',
+            })
+            relatedInformation.push(fileXML)
+        }
     }
+
+    const allRelatedInformation = relatedInformation.join('\n')
+    const messageInfo = xmlBuilder.build({
+        message: message.text || '',
+        allRelatedInformation: allRelatedInformation,
+    })
+    return messageInfo
 }
 
 async function fetchThreadMessages(
     client: WebClient,
     channelId: string,
     threadTs: string
 ): Promise<string | null> {
-    const allMessages = []
+    const allMessages: MessageElement[] = []
     try {
         let response = await client.conversations.replies({ channel: channelId, ts: threadTs })
-        let messages = response.messages as any[]
-        allMessages.push(...messages)
+        if (response.messages) {
+            allMessages.push(...response.messages)
+        }
         while (response.has_more) {
             const nextCursor = response.response_metadata?.next_cursor
             response = await client.conversations.replies({
                 channel: channelId,
                 ts: threadTs,
                 cursor: nextCursor,
             })
-            messages = response.messages as any[]
-            allMessages.push(...messages)
+            if (response.messages) {
+                allMessages.push(...response.messages)
+            }
         }
-        return allMessages?.map(msg => msg.text).join('\n\n')
+        return allMessages?.map(msg => getMessageInformation(msg)).join('\n\n')
     } catch (error) {
         return null
     }
 }
 
+function extractThreadTs(permalink: string | undefined): string | null {
+    if (!permalink) {
+        return null
+    }
+    const url = new URL(permalink)
+    const queryParams = new URLSearchParams(url.search)
+    const threadTs = queryParams.get('thread_ts')
+    return threadTs
+}
+
+function extractChannelId(slackUrl: string | undefined): string | null {
+    if (!slackUrl) {
+        return null
+    }
+    const pattern = /archives\/([A-Z0-9]+)\//
+    const match = slackUrl.match(pattern)
+    if (match) {
+        return match[1]
+    }
+    return null
+}
+
+// Function to search messages using slack api
+async function searchQuery(
+    client: WebClient,
+    query: string,
+    numResults = 5
+): Promise<Match[] | undefined> {
+    try {
+        const result = await client.search.messages({ query: query, count: numResults })
+        return result?.messages?.matches
+    } catch (error) {
+        return undefined
+    }
+}
+
+async function getContextFromSlackSearchApi(
+    client: WebClient,
+    channelId: string | undefined,
+    query: string,
+    channelLimit = 50,
+    workspaceLimit = 50
+) {
+    if (channelId) {
+        const [channelSearchResults, workspaceSearchResults] = await Promise.all([
+            searchQuery(client, `in:<#${channelId}> ${query}`, channelLimit),
+            searchQuery(client, query, workspaceLimit),
+        ])
```

Comment:
Is there a benefit of using `Promise.all()` here instead of two sequential `await searchQuery()` calls for `channelSearchResults` & `workspaceSearchResults`?

Reply:
promise.all will fetch them in parallel whereas sequantially will wait for the first one finish.

Reply:
Hmm 🤔

I don't know if `Promise.all()` * *technically* * runs in parallel but even if it did, we might be dealing with a relatively minor tradeoff b/w performance & readability.

anyway, just a nit

Reply:
Yeah, no doubts on the I/O delays. I couldn't find a definitive reference on `Promise.all()` running in parallel, but it certainly seems to [speed up](https://gist.github.com/tinvaan/822f6ba2ece54d95e179d90b0223863e) a simple script that performs a `setTimeout()`.

Looks good to me 👍🏼